### PR TITLE
Fixed password change issue from profile page

### PR DESCRIPTION
**Pull request by passionInfinite**
#204  and #369 are resolved in this Pull Request.

### CODE REVIEW

File: app/app/Http/Controllers/ChangePasswordController.php [hunk 1]
```diff
@@ -32,36 +33,26 @@ public function post(Request $request)
...
-			{
-				if ( !BaseUser::authenticate($user->email, $request->input('current_password')) )
-				{
-					$validator->errors()->add('current_password', 'Current password not matched');
-					$failing = true;
-				}
+			$validator = Validator::make(
+			    $request->all(), [
+			        'current_password' => 'required',
+                    'new_password' => 'required',
```
Comment (joshi1983) [inline app/app/Http/Controllers/ChangePasswordController.php]: Fix the indentation here.  It is really messed up.  current_password and new_password should line up.

This could be fixed by just undoing your changes to the $validation_rules variable.

Check your editor's tab and space preferences.

File: app/resources/views/pages/profile/profile.blade.php [hunk 2]
```diff
@@ -40,6 +39,11 @@
 		</form>
  	</div>
     <div class="col-md-9 col-sm-8 col-xs-12">
+		@if (Session::has('message'))
+			<div class="alert alert-success alert-dismissible" role="alert" style="margin-top: 16px">
```
Comment (joshi1983) [inline app/resources/views/pages/profile/profile.blade.php]: Move inline styles like your "margin-top" to the appropriate SCSS file.  I think resources/assets/sass/_profile.scss is right for this.

File: app/resources/views/pages/profile/profile.blade.php [hunk 3]
```diff
@@ -1,21 +1,20 @@
...
+
+	<link href="/css/jquery/jquery-ui.css" rel="stylesheet" type="text/css">
+	<script src="/css/jquery/external/jquery/jquery.js"></script>
+	<script src="/css/jquery/jquery-ui.js"></script>
+	<script src="/js/profile.js"></script>
+	<script src="/js/profile_save_button.js"></script>
+	<script src="/js/profile_save_prompt.js"></script>
+	<script src="/js/utils.js"></script>
+	<script src="/js/question_explanation.js"></script>
+@sto
```
Comment (joshi1983) [inline app/resources/views/pages/profile/profile.blade.php]: Fix the @stop missing the 'p' and remove these indentation changes.  The indentation changes here make it difficult to annotate who added each library and why because 'git blame' will show the same commit for all of these lines.

File: app/app/Http/Controllers/ChangePasswordController.php [hunk 4]
```diff
@@ -32,36 +33,26 @@ public function post(Request $request)
...
-			$validator = Validator::make(Input::all(), $validation_rules);
-			$failing = $validator->fails();
-			if (!$failing && !empty($user->password_hash))
-			{
-				if ( !BaseUser::authenticate($user->email, $request->input('current_password')) )
-				{
-					$validator->errors()->add('current_password', 'Current password not matched');
-					$failing = true;
-				}
+			$validator = Validator::make(
```
Comment (joshi1983) [inline app/app/Http/Controllers/ChangePasswordController.php]: Why not leave the $validation_rules variable unchanged?

File: app/app/Http/Controllers/ChangePasswordController.php [hunk 5]
```diff
@@ -32,36 +33,26 @@ public function post(Request $request)
...
+			    $request->all(), [
+			        'current_password' => 'required',
+                    'new_password' => 'required',
+                    'password_confirm' => 'required|same:new_password'
+                ]
+            );
+			if ($validator->fails()) {
+			    return redirect()->to('change-password')->withErrors($validator->errors())->withInput();
+            } else {
+			    if (Hash::check($request->get('current_password'), $user->password_hash)) {
```
Comment (joshi1983) [inline app/app/Http/Controllers/ChangePasswordController.php]: BaseUser::authenticate does the Hash::check in a central place so why did you replace with Hash class directly here?

File: app/public/fonts/bootstrap/glyphicons-halflings-regular.svg [hunk 6]
```diff
@@ -0,0 +1,288 @@
+<?xml version="1.0" standalone="no"?>
```
Comment (joshi1983) [inline app/public/fonts/bootstrap/glyphicons-halflings-regular.svg]: Why did you change the bootstrap font graphic files in this commit?  That seems unrelated to the password change issue.  I noticed you had these files changed in your practice pull request too.   Was this a mistake?

File: app/app/Http/Controllers/ChangePasswordController.php (hunk 5 above)
Comment (passionInfinite) [reply app/app/Http/Controllers/ChangePasswordController.php]: Because I just wanted to check the password don't wanted user to login.

File: app/app/Http/Controllers/ChangePasswordController.php (hunk 4 above)
Comment (passionInfinite) [reply app/app/Http/Controllers/ChangePasswordController.php]: No Problem, I will update it

File: app/app/Http/Controllers/ChangePasswordController.php (hunk 1 above)
Comment (passionInfinite) [reply app/app/Http/Controllers/ChangePasswordController.php]: Don't know why it showed like this here but in my editor it is showing perfect.

File: app/app/Http/Controllers/ChangePasswordController.php (hunk 5 above)
Comment (joshi1983) [reply app/app/Http/Controllers/ChangePasswordController.php]: BaseUser::authenticate doesn't log a user in.  It just returns true or false indicating if the email and password matched a user.  authenticate doesn't have any side effects.

signin is the method you're talking about which saves that the user signed in.

File: app/public/fonts/bootstrap/glyphicons-halflings-regular.svg (hunk 6 above)
Comment (passionInfinite) [reply app/public/fonts/bootstrap/glyphicons-halflings-regular.svg]: Actually, whenever I work with the new branch it doesn't have public/fonts folder so it creates them. once it is merged with the master it won't happen again.

**Comment by passionInfinite** [comment]
@joshi1983 Update the PR.

File: app/public/fonts/bootstrap/glyphicons-halflings-regular.svg (hunk 6 above)
Comment (joshi1983) [reply app/public/fonts/bootstrap/glyphicons-halflings-regular.svg]: gulp generates those files, though.  Tracking them is like tracking the app.css file that all of our scss is converted into.

If you're bothered by running gulp when changing branches, you could either keep the files untracked/uncommitted too.  You may also like using 'gulp watch' instead of just 'gulp' since 'gulp watch' will automatically run when changed are saved to input files.

I merged a change tonight to ignore files in public/fonts/bootstrap to help with the issue you ran into and clarify why we don't want to track those files.

Please delete your public/fonts/bootstrap directory and squash or fixup that into your other commit.  Combining the changes into a single commit will prevent a conflict between other developers having those files untracked and one of your commits adding the same files.

File: app/resources/assets/sass/_profile.scss [hunk 7]
```diff
@@ -256,3 +256,7 @@
 		}
 	}
 }
+
+.margin-top-12 {
```
[the remark this answers is not in the document]
Comment (passionInfinite) [reply app/resources/assets/sass/_profile.scss]: Lets make this as common so that it can be used anywhere when required.

**Comment by passionInfinite** [comment]
@joshi1983  Updated the PR.

File: app/app/Http/Controllers/ChangePasswordController.php [hunk 8]
```diff
@@ -32,36 +33,25 @@ public function post(Request $request)
...
+			$validationRules = [
+                'current_password' => 'required',
+                'new_password' => 'required',
+                'password_confirm' => 'required|same:new_password'
+            ];
+			$validator = Validator::make($request->all(), $validationRules);
+			if ($validator->fails()) {
+			    return redirect()->to('change-password')->withErrors($validator->errors())->withInput();
+            } else {
+			    if (Hash::check($request->get('current_password'), $user->password_hash)) {
```
[the remark this answers is not in the document]
Comment (joshi1983) [reply app/app/Http/Controllers/ChangePasswordController.php]: You don't get why we're asking for the current password despite being signed in?

This can explain it: https://security.stackexchange.com/questions/24291/why-do-we-ask-for-a-users-existing-password-when-changing-their-password

[the remark this answers is not in the document]
Comment (joshi1983) [reply app/app/Http/Controllers/ChangePasswordController.php]: It'll be a little less extra code.  No need for the "use Illuminate\Support\Facades\Hash" at the top.  It is also to keep the hash checks in a single place.  I admit these aren't huge reasons but there's no bigger reason to use Hash::check either so just keep using BaseUser::authenticate as I requested.

**Comment by joshi1983** [comment]
This pull request had some remaining tasks for a few months so another pull request was made to include the desired work from this one.  That new pull request is at #704.

I'm closing this as the other pull request was merged with the desirable changes.
